### ISSUE TITLE
[Feature request] Improve OpenShift integration with Gradle and support for MSSQL

### ISSUE THREAD

**Issue by Spengreb** [jhipster/generator-jhipster#6766]
##### **Overview of the issue**
The current implementation of OpenShift integration does not support any Gradle tasks and depends on the OpenShift CLI. It also seems to favor micro-service spring boot applications, creating an ocp/ the directory above the project root. Furthermore it also does not support MSSQL (as there is no stable Docker version).  

##### **Motivation for or Use Case**

- I wanted to deploy to OpenShift through GitLab CI and so I wanted to deploy via Gradle. 
- I use a Monolithic version and so I was not satisfied with what was generated (and was a bit confused at first). 
- I also use MSSQL with the paradigm of each branch I create having its own database in a single MSSQL instance. 

##### **Suggest a Fix**
I created a Gradle plugin with opinionated tasks for deploying to OpenShift. https://gitlab.com/hendosdo/gradle-openshift-tasks-plugin

This will start builds on Openshift and send deployments to it. This is done by modifying the deployment.yml a bit for updating application name and updating database location. 

I also have a local version of JHipster which will generate files properly in places the Plugin expects. I also retained existing functionality. It asks a few more questions as a result, like if mssql is selected it will ask for username, password, and URL. 

I will submit a PR if there is interest in this.

##### **JHipster configuration**
Running yo jhipster:openshift will create appropriate files, it checks the yo-rc.json file for existing configuration

- [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by jdubois** [comment]
- The MSSQL image we use is `'microsoft/mssql-server-linux:latest'` so I don't understand why you say we don't support it/we don't have a stable version? Anyway, in production, I wouldn't recommend running your database in Docker, but to use a managed service.
- I also don't understand your 3rd point - is it worrying you that we have one database per microservice? That's the whole point of doing microservices: if you have one single database, then you just have a distributed monolith. Note that if you want to do this, it isn't very hard, there is just one config key in `application-*.yml` to modify.
- I'm not sure what Gradle does here: our OpenShift support should work both with Maven and Gradle, and what I read at http://www.jhipster.tech/openshift/ looks good -> I'm not an expert here, @srinivasavasu is the boss on this, so I'm affecting the ticket to him, and hope he can answer us

**Comment by deepu105** [comment]
@Spengreb 
For your point 1, it's a new feature you request which I don't think we will support. Our workflow is to use the sub-generators for deployment. If you want you could write a tip on the doc site explaining how to do it with your plugin. If I misunderstood your point please do clarify

You can submit a PR for the other stuff if you wish and we will review to see if it is good to go. And of course we will wait for @srinivasa-vasu to take a look at your PR as well

**Comment by Spengreb** [comment]
Thanks for your responses.

I think there is a bit of confusion here. 

@jdubois 
- I wasn't aware that the mssql docker version had left beta already! But even still this is not supported in the openshift sub-generator. The Openshift sub-generator does not add another pod with this docker image, like it does for other DB engines. In fact the sub-generator will hang at the moment if MSSQL is selected in JHipster. We were just trying out the hosted RDS service which also isn't supported. The OpenShift sub-generator assumes you want the DB as a pod in OpenShift always.   
- It is not worrying me about one database per application for microservices, I agree with that. What I'm not happy about is the OpenShift sub-generator being geared towards microservices, i.e. regardless of what you answer in the OpenShift sub-generator prompts (microservice or monolithic) it will always assume microservice as this is part of the DockerPrompts bit and is not used in the Openshift sub-gen. In my local version if you select monolithic it will create DeploymentConfigs in your application, not in the directory above. 
- The sub-generator will work with both maven and gradle, what my plugin does is add tasks for openshift **deployment** in gradle. It adds 4 tasks, a package task which takes in Dockerfiles + War, a task for sending that off to OpenShift to be built into a pod, and then a task for sending the DeploymentConfig off to openshift to start it off. The last task does all this in order. I can't add these tasks to Maven as I don't use it personally at all. 

So to summarize, I have altered the OpenShift sub-generator to create a single DeploymentConfig file if monolithic is selected, it will add a openshift.gradle file into the java project which has configurations for my Openshift Gradle **deployment** plugin, and to add lines into the DeploymentConfig if MSSQL is selected and assumes a hosted service.

I realize that last bit may not be exactly what you're looking for but its what my assumption was before I knew about the MSSQL docker image being out of Beta. 

Hope this clears things up. Will submit a PR for you guys to have a look

**Comment by deepu105** [comment]
closing as PR closed
